addons/adyen_marketpay/models/adyen_account.py

```diff
@@ -0,0 +1,640 @@
+    payout_ids = fields.One2many('adyen.payout', 'adyen_account_id', string='Payouts')
+    shareholder_ids = fields.One2many('adyen.shareholder', 'adyen_account_id', string='Shareholders')
+    bank_account_ids = fields.One2many('adyen.bank.account', 'adyen_account_id', string='Bank Accounts')
+    transaction_ids = fields.One2many('adyen.transaction', 'adyen_account_id', string='Transactions')
+    transactions_count = fields.Integer(compute='_compute_transactions_count')
+
+    is_business = fields.Boolean('Is a business', required=True)
+
+    # Contact Info
+    full_name = fields.Char(compute='_compute_full_name')
+    email = fields.Char('Email', required=True)
+    phone_number = fields.Char('Phone Number', required=True)
+
+    # Individual
+    first_name = fields.Char('First Name')
+    last_name = fields.Char('Last Name')
+    date_of_birth = fields.Date('Date of birth')
+    document_number = fields.Char('ID Number', help="The type of ID Number required depends on the country:\nUS: Social Security Number (9 digits or last 4 digits)\nCanada: Social Insurance Number\nItaly: Codice fiscale\nAustralia: Document Number")
+    document_type = fields.Selection(string='Document Type', selection=[
+        ('ID', 'ID'),
+        ('PASSPORT', 'Passport'),
+        ('VISA', 'Visa'),
+        ('DRIVINGLICENSE', 'Driving license'),
+    ])
+
+    # Business
+    legal_business_name = fields.Char('Legal Business Name')
+    doing_business_as = fields.Char('Doing Business As')
+    registration_number = fields.Char('Registration Number')
+
+    # KYC
+    kyc_status = fields.Selection(string='KYC Status', selection=[
+        ('awaiting_data', 'Data to provide'),
+        ('pending', 'Waiting for validation'),
+        ('passed', 'Confirmed'),
+        ('failed', 'Failed'),
+    ], required=True, default='awaiting_data')
+    kyc_error_message = fields.Char('KYC Error Message', readonly=True)
+
+    _sql_constraints = [
+        ('adyen_uuid_uniq', 'UNIQUE(adyen_uuid)', 'Adyen UUID should be unique'),
+    ]
+
+    @api.depends('transaction_ids')
+    def _compute_transactions_count(self):
+        for adyen_account_id in self:
+            adyen_account_id.transactions_count = len(adyen_account_id.transaction_ids)
+
+    @api.depends('first_name', 'last_name', 'legal_business_name')
+    def _compute_full_name(self):
+        for adyen_account_id in self:
+            if adyen_account_id.is_business:
+                adyen_account_id.full_name = adyen_account_id.legal_business_name
+            else:
+                adyen_account_id.full_name = "%s %s" % (adyen_account_id.first_name, adyen_account_id.last_name)
+
+    @api.model
+    def create(self, values):
+        adyen_account_id = super(AdyenAccount, self).create(values)
+        self.env.company.adyen_account_id = adyen_account_id.id
+
+        # Create account on odoo.com, proxy and Adyen
+        response = adyen_account_id._adyen_rpc('create_account_holder', {
```

Comment:
Can't we do all rpc before calling super and setting the account on the company, like that you'll be able to set the adyen_uuid and proxy_token in vals dict before creation.

Maybe the creation of payout, should still be done after the super

Reply:
If something fails during the call to super (e.g due to an SQL constraint) the user would be created in odoo.com, the proxy, and Adyen but not in the customer DB. I wanted to process as much data as I could before creating the submerchant in our Adyen account to avoid this kind of problem.

Reply:
ok :)